docs/transformations/CustomTransformation.md

```diff
@@ -0,0 +1,212 @@
+| String (**enclosed in single quotes**, e.g., 'Transformed text')                    | TEXT, ENUM, JSON, CHAR, LONGTEXT, MEDIUMTEXT, SET, TINYTEXT, VARCHAR             |
+| Binary String                                                                       | BLOB, BINARY, BIT, LONGBLOB, MEDIUMBLOB, TINYBLOB, VARBINARY                     |
+| String (Format: yyyy-MM-dd **enclosed in single quotes**, e.g., '1995-01-13')       | DATE                                                                             |
+| String (e.g., 2024-01-01T12:34:56Z)                                                 | DATETIME, TIMESTAMP                                                              |
+| String (**enclosed in single quotes**)                                              | [Spatial Datatypes](https://dev.mysql.com/doc/refman/8.0/en/spatial-type-overview.html) |
+
+{: .highlight }
+Please refer to the sample implementation of **toSourceRow** for most MySQL datatype columns [here](https://github.com/GoogleCloudPlatform/DataflowTemplates/blob/main/v2/spanner-custom-shard/src/main/java/com/custom/CustomTransformationWithShardForIT.java#L145).
+
+## Steps to implement custom transformation
+1. Checkout the dataflow code from [github](https://github.com/GoogleCloudPlatform/DataflowTemplates)
+    ```
+    git clone https://github.com/GoogleCloudPlatform/DataflowTemplates.git
+    ```
+2. Update the logic in [CustomTransformationFetcher](https://github.com/GoogleCloudPlatform/DataflowTemplates/blob/main/v2/spanner-custom-shard/src/main/java/com/custom/CustomTransformationFetcher.java) class in `v2/spanner-custom-shard`. In case of only forward migration only updating the toSpannerRow is sufficient and for toSourceRow just use the below sample:
+    ```
+    @Override
+    public MigrationTransformationResponse toSourceRow(MigrationTransformationRequest request throws InvalidTransformationException {
+    return new MigrationTransformationResponse(null, false);
+    }
+    ```
+
+    Similarly, in case of only reverse replication updating the toSourceRow is sufficient and for toSpannerRow just use the below sample:
+    ```
+    @Override
+    public MigrationTransformationResponse toSpannerRow(MigrationTransformationRequest request throws InvalidTransformationException {
+    return new MigrationTransformationResponse(null, false);
+    }
+    ```
+3. If any custom parameters are needed in the custom transformation logic, they can be passed to the *init* method of the custom class. The *init* method is invoked once per worker setup.
+3. Please test the modified code by writing unit and cross functional tests in [CustomTransformationFetcherTest.java](https://github.com/GoogleCloudPlatform/DataflowTemplates/blob/main/v2/spanner-custom-shard/src/test/java/com/custom/CustomTransformationFetcherTest.java) 
+4. Build the `spanner-custom-shard` module by running the below commands:
+    ```
+    cd v2/spanner-custom-shard
+    mvn install
```

Comment:
will the user need to do an mvn install on the parent directory for htis to work? I remember one had to do this one-time mvn install (Even if it fails its okay as long as a certain module gets built) on parent before building a template. Not sure if it is a requirement on this as well.

Reply:
I have tested this on a fresh cloned repo, it works. It doesn't have dependency on parent directories that use to cause this issue in the past